[sparse] Autograd get_indices/values and sparse_coo ctor

ssnl opened the pull request:
TODO: docs

Closes #11232

ssnl committed:
Get rid of HARDCODED_DIFFERENTIABLE_OUTPUTS. Replace it with
`output_differentiability` in derivatives.yaml. Also relax the
check that gradient formulas need to use all grad outputs. It is
well possible that to compute a particular grad_input[i], only
part of all grad_ourputs are needed.

add sparse get_values and make it back-prop-able

Make get_values back-prop-able

make indices and values view functions

Make all sparse_coo ctors dispatch to a native function,
_sparse_new_with_dims_and_tensor. Remove the dispatch mechaism
on native_* native ctors, e.g., native_sparse_coo_tensor. Now
all the code lives in functions like sparse_coo_tensor.

Make sparse coo ctor a view function

Make _newFlattenedIndices a native function
Implement sparse_constructor_backward

Get rid of NNZ optimization
Move native/sparse/SparseUtils.h to SparseTensorUtils.h

add getter docs

make _set_coalesced a native fn and call it _coalesced_

sparseDims -> sparse_dim; denseDims -> dense_dim

update test_print expect because I fixed _indices output to not have grad_fn now

infer type first

get_indices -> indices; get_values -> values

purge options from sparse_coo_tensor with indices and values tensors

Fix coalesced tests; update prints; use type dispatch for size only ctor

Update note; support nondiff views; update prints

workaround for sparse views and inplace ops

SsnL commented:
@pytorchbot retest this please

ezyang commented on tools/autograd/derivatives.yaml:
```diff
@@ -8,6 +8,12 @@
 #     Note that a single gradient entry can specify the gradient
 #     formula for multiple input names, by specifying a key
 #     "input1, input2" (see atan2 for an example).
+#   - Optional entry with key 'output_differentiability' and value a list of the
```
So... this means that we should forbid people from declaring native functions whose has an argument name `output_differentiability`, right? :)

SsnL replied:
Technically no, because each entry in this yaml has `name: native_fn_declaration`. The name of the function isn't a key. But I can't think of a scenario where such a function name would make sense. :D

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -283,11 +283,12 @@ def is_differentiable(arg):
-        for i in hardcoded_diff:
-            differentiable_outputs.append(candidate_differentiable_outputs[i])
+        output_differentiability = func.get('output_differentiability')
+        for differntiable, output in zip(output_differentiability, candidate_differentiable_outputs):
```
nit: mispelled

SsnL commented:
@ezyang Yes, I agree. Also sorry about the commit structure being a bit messed up... I changed design in the middle of doing this patch.

SsnL commented:
@ezyang re: method name. Didn't have a particular reason. I like `indices`. It's shorter and more like an attribute, I'll probably go by that.

yf225 commented:
@SsnL are we ready to merge this PR after rebase? :)

ezyang commented:
Bump on this PR!

SsnL commented:
@ezyang Heard you! I'm rebasing rn!

SsnL commented:
sigh. I can't reproduce these errors locally.

ezyang commented:
Try building in the Docker? :) https://github.com/pytorch/ossci-job-dsl

SsnL commented:
@ezyang Yep I plan to do that tomorrow. :)

SsnL commented:
need to implement nondifferentiable view first. so I will first tackle https://github.com/pytorch/pytorch/issues/11390

ssnl committed:
View op outputs are not registered as views when !GradMode::enabled()

ssnl committed:
potential_history_tracking -> potentially_tracks_history

ssnl committed:
make the note clearer

ssnl committed:
update note

SsnL commented:
This is now stacked on top of #12502

ssnl committed:
more comments

ssnl committed:
diff and nondiff views

ssnl committed:
more comments

ssnl committed:
rename note

ssnl committed:
fix typos

ssnl committed:
typo

ssnl committed:
Use function variant with option dispatch; favor options dispatch over tensor dispatch

ssnl committed:
arg checking and de-duplicate code

ssnl committed:
revert grad mode stuff

ssnl committed:
Make _values non-differentiable; Add note on why
Add has_* for TensorOptions
Fix Python sparse_coo_tensor entry
Fix a CUDA coalesce error; add tests

ssnl committed:
update expect because _values is nondifferentiable

ssnl committed:
Fix narrow_copy_sparse compilation

ssnl committed:
Revert unintended derivatives.yaml change;

ssnl committed:
revert no_grad viewer test since that part of the change is not getting in this time

ssnl committed:
Fix test_numba_integration may try to produce impossible sparse tensor

SsnL commented:
@pytorchbot retest this please

ssnl committed:
fix numba test & improve argcheck

ssnl committed:
skip autograd ifRocm

ssnl committed:
debug tensorFromBlob with variable type

SsnL commented:
i'm using github to sync code between local and remote to debug. so i'll close this PR to save CI runs, and reopen afterwards
